Issue #17841: Add GoogleMethodNameCheck to fix method name false nega…

vivek-0509 opened the pull request:
part of Issue: #17841
Issue: #18420 

## Description

Introduces a new `GoogleMethodName` check that properly validates method names according to the Google Java Style Guide, fixing false-negatives where underscores between letters and digits were not being flagged.

### Problem

The previous `MethodName` check failed to detect invalid names like:
- `gradle_9_5_1()` - underscore between letter 'e' and digit '9'
- `jdk_9_0_392()` - underscore between letter 'k' and digit '9'

### Solution

Created a dedicated `GoogleMethodName` check that:
- Has no configurable properties
- Uses multi-pattern validation instead of one complex regex
- Handles `@Test` methods separately (underscores allowed)
- Properly validates numbering suffixes (`guava33_4_5` is valid)

### Validation Rules

**Regular Methods:**
- Must be in lowerCamelCase
- Underscores only between adjacent digits (`_4_5`)
- Not between letter and digit (`gradle_9` invalid)

**Test Methods (`@Test`, `@ParameterizedTest`, `@RepeatedTest`):**
- Underscores allowed to separate components
- Each segment must be lowerCamelCase
- Each segment must be at least 2 characters
- Segments cannot start with single lowercase followed by uppercase (e.g., fO)

**Override Methods:**
- Methods with `@Override` annotation are skipped (no validation)

### Breaking Changes
None. This check is backward compatible with the previous [MethodName](cci:2://file:///Users/vivek/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/MethodNameCheck.java:40:0-160:1) regex behavior.

Diff Regression config: https://gist.githubusercontent.com/vivek-0509/3c0bc74332fbdb3a6ce4aef1c43d2f6b/raw/a7d36ba756226bac0e01e9148dc8bd7f3016ec96/base_config.xml

Diff Regression patch config: https://gist.githubusercontent.com/vivek-0509/a2777cd3e2a86dc84c549268d6651aa0/raw/6490530b3efb4213b7a2421dace6f157c7b264a2/patch_config.xml

Contirbution repo PR:  https://github.com/checkstyle/contribution/pull/1000

vivek-0509 commented:
The CI Failures are unrelated

vivek-0509 commented:
GitHub, generate report

vivek-0509 commented:
GitHub, generate report

vivek-0509 commented:
GitHub, generate report

vivek-0509 commented:
GitHub, generate report

vivek-0509 commented:
Hey @romani,
Here is the analysis of the report

# Bug Found (Will Fix)

**Issue**: Test methods with numbering suffix like `test3_1` are incorrectly rejected.

**Cause**: The implementation doesn't strip numbering suffix before validating test methods (unlike regular methods).

**Examples**:
- `test3_1` → **Rejected** (Should be valid - 3_1 = adjacent numbers)
- `testFoo_1_2` → **Rejected** (correctly invalid)

**Fix**: Will add numbering suffix stripping + `INVALID_UNDERSCORE_PATTERN` check for test methods.

---

## Design Questions - Need Your Input

### 1. Single lowercase + Uppercase pattern (`xAmbar`, `pLevel`, `eMeSinE`)

Original `MethodName` regex has `(?![a-z][A-Z])` which rejects names starting with single lowercase followed by uppercase.

**Examples**:
- `xAmbar` → **Rejected** (arguably valid lowerCamelCase: x + Ambar)
- `xPathRuleQueryNeedsInitialization` → **Rejected**

**Question**: Keep for backward compatibility, or fix as **bug**?

### 2. Minimum 2-character segment requirement for test methods

Currently rejects single lowercase character segments:

**Examples**:
- `t_name` → **Rejected** (segment `t` = 1 char)
- `test_a` → **Rejected** (segment `a` = 1 char)
- `test_foo` → **Valid** (both segments ≥ 2 chars)

**Real example**:
```
@Test
void can_be_used_in_a_collector() {} //  Rejected (segment a fails)
```

**Question**: Keep 2-char minimum or allow single-char segments?

### 3. Test methods without @Test annotation

**Examples**:
- `void test_ctor_shouldThrow_whenPidIsMissing()` → **REJECTED** (no annotation)
- `@Test void test_ctor_shouldThrow_whenPidIsMissing()` → **VALID**

**Supported annotations**: `@Test`, `@ParameterizedTest`, `@RepeatedTest`

**Question**: Should test-style method names without annotations be rejected, as Google Style only mentions JUnit tests?

vivek-0509 commented:
Hi @Zopsss ,

Before you review the code, I'd like to clarify my understanding of the expected rules for GoogleMethodNameCheck , as Google Style Guide is somewhat ambiguous in places.

For regular methods (without `@Test`), my implementation requires lowerCamelCase names where underscores are only allowed between adjacent numbers (like guava33_4_5). Names like gradle_9_5_1 with letter_digit underscores are flagged as invalid. Single letter names like `f` and patterns like fO are also rejected.

For test methods (with `@Test`, `@ParameterizedTest`, `@RepeatedTest`), underscores are allowed to separate logical components, with each component being lowerCamelCase. So [transferMoney_deductsFromSource](cci:1://file:///Users/vivek/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/googlemethodname/InputGoogleMethodNameValidTest.java:14:4-16:5) is valid, but [test_1](cci:1://file:///Users/vivek/checkstyle/src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java:92:2-93:18) is invalid because `1` is not lowerCamelCase and also it violates the no underscore between just number and digit. The numbering suffix rule also applies, so [test3_1](cci:1://file:///Users/vivek/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/googlemethodname/InputGoogleMethodNameValidTest.java:51:4-52:21) is valid (adjacent numbers), but [testFoo_1_2](cci:1://file:///Users/vivek/checkstyle/src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java:98:2-99:23) is invalid (letter_digit pattern). For test methods like [solve6x6_returnsTrue](cci:1://file:///Users/vivek/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/googlemethodname/InputGoogleMethodNameValidTest.java:18:4-20:5) and [solve6x6_noSolution_returnsFalse](cci:1://file:///Users/vivek/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/googlemethodname/InputGoogleMethodNameValidTest.java:22:4-24:5), my implementation treats them as valid since each component ([solve6x6](cci:1://file:///Users/vivek/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/googlemethodname/InputGoogleMethodNameValidTest.java:18:4-20:5), [returnsTrue](cci:1://file:///Users/vivek/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/googlemethodname/InputGoogleMethodNameValidTest.java:18:4-20:5), [noSolution](cci:1://file:///Users/vivek/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/googlemethodname/InputGoogleMethodNameValidTest.java:22:4-24:5), [returnsFalse](cci:1://file:///Users/vivek/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/googlemethodname/InputGoogleMethodNameValidTest.java:22:4-24:5)) is lowerCamelCase and the underscore separates these components rather than being between a digit and letter within a component.

One thing I'm unclear about: Google Style says test components should be "lowerCamelCase". Strictly speaking, lowerCamelCase means multiple words joined with capitalization (like [transferMoney](cci:1://file:///Users/vivek/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/googlemethodname/InputGoogleMethodNameValidTest.java:14:4-16:5)). Should single-word components like [test_foo](cci:1://file:///Users/vivek/checkstyle/src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName.java:95:2-96:22) be valid, or should we require multi-word lowerCamelCase like [transferMoney_deductsFromSource](cci:1://file:///Users/vivek/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/googlemethodname/InputGoogleMethodNameValidTest.java:14:4-16:5)?

If any of my expectations don't align with what's actually required, please let me know and I'll adjust my implementation before you do a full code review. This should save time .

Zopsss commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/GoogleMethodNameCheck.java:
```diff
@@ -0,0 +1,194 @@
+ * <p>
+ * This check has no configurable properties. It is specifically designed to
+ * enforce Google Java Style and handles test method annotations internally.
+ * </p>
```
```
 * <p>
 * This check has no configurable properties. It is specifically designed to
 * enforce Google Java Style rules: 5.2.3 Method Names and 5.3 Camel case: defined
 * </p>
```

the rule names should be link.

Zopsss commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/GoogleMethodNameCheck.java:
```diff
@@ -0,0 +1,194 @@
+            }
+            else {
+                validateRegularMethodName(nameAst, methodName);
+            }
```
```suggestion
            if (!hasTestAnnotation(ast)) {
                validateRegularMethodName(nameAst, methodName);
            }
            else {
                validateTestMethodName(nameAst, methodName);
            }
```

More optimized because there will be a lot of regular methods compare to test methods.

Zopsss commented on src/main/resources/com/puppycrawl/tools/checkstyle/checks/naming/messages.properties:
```diff
@@ -1,4 +1,5 @@
 abbreviation.as.word=Abbreviation in name ''{0}'' must contain no more than ''{1}'' consecutive capital letters.
+google.method.name=Method name ''{0}'' is not valid per Google Java Style Guide.
```
this message does not explain why the method name is invalid. Instead we should keep different messages for the different conditions we hit to easily explain reason behind violation to the user.

Zopsss commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/GoogleMethodNameCheck.java:
```diff
@@ -0,0 +1,194 @@
+                log(nameAst, MSG_KEY, methodName);
+            }
+        }
+    }
```
the `test_Test` is checked at `if (!REGULAR_METHOD_NAME_PATTERN.matcher(nameWithoutNumberingSuffix).matches()) {` even though it is an invalid underscore pattern.

It should be checked at the INVALID_UNDERSCORE_PATTERN

vivek-0509 commented:
Hi @Zopsss, thanks for the review 
I’ve created an issue to clarify the expected scope and rules for this check: https://github.com/checkstyle/checkstyle/issues/18420

Would it be acceptable if I apply all the requested changes once we align on what exactly this check should cover and the set of rules to be enforced for this check?

Zopsss commented:
GitHub, generate report

vivek-0509 commented:
GitHub, generate report

romani reviewed (changes requested):
oh, you already updated all google style 
lets still split update into two commits.
first commit - to add new Check
second commit to add to google style new module (reference two different issues in commit messages).

we might be lucky to merge all at ones, but if there will be complications, I will demand two separate PRs, for now lets try all atones. 

items:

romani commented on src/it/resources/com/google/checkstyle/test/chapter5naming/rule53camelcase/InputUnderscoreUsedInNames.java:
```diff
@@ -148,7 +161,9 @@ void testAddAll_nonEmptyList() {}
     @Test
-    void testAddAll_NonEmptyList() {} // violation, uppercase letter after _, 'pattern'
+    void testAddAll_NonEmptyList() {
+      // violation above, 'Test method name .* is not valid, each segment must be lowerCamelCase.*'
```
please keep `{}` and put comments above annotation.
do same for all others

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/GoogleMethodNameCheck.java:
```diff
@@ -0,0 +1,227 @@
+ *
+ * <p>
+ * This check has no configurable properties. It is specifically designed to
+ * enforce Google Java Style rules:
```
please remove this paragrath

romani reviewed (changes requested):
items to improve:

vivek-0509 commented on src/it/resources/com/google/checkstyle/test/chapter5naming/rule523methodnames/InputMethodName2.java:
```diff
@@ -8,30 +8,31 @@ public class InputMethodName2 {
 
-  @Test
+  @Test // violation below, 'Test method name .* has invalid underscore usage.*'
   public void solve6x6_returnsTrue() {}
```
since the underscore is present between a digit and letter , https://github.com/checkstyle/checkstyle/issues/18420#issuecomment-3692865811

vivek-0509 replied:
I have a small request, if possible. Since the reviews have been a bit spaced out, I was wondering if we could review this one more continuously for a short period. I believe we’re quite close, and with a few focused iterations we could get it merged soon.
I’ll make sure to respond quickly to any feedback.

vivek-0509 commented on config/jsoref-spellchecker/whitelist.words:
```diff
@@ -1009,6 +1010,7 @@ onestatementperline
 onetoplevel
 onetoplevelclass
 onevariableperline
+oo
```
done

vivek-0509 commented on src/it/resources/com/google/checkstyle/test/chapter5naming/rule53camelcase/InputCamelCaseDefined.java:
```diff
@@ -45,13 +45,13 @@ void YoutubeImporter() {}
 
             void YoutubeImporter() {}
-            // violation above 'Method name 'YoutubeImporter' must match pattern'
+            // violation above, ''YoutubeImporter' must be lowerCamelCase.*'
```
@romani, Could you please confirm if this violation messages works ??


```
google.method.name.format.regular=Method name ''{0}'' must start with a lowercase letter, min 2 chars, avoid single lowercase letter followed by uppercase.

google.method.name.format.test=Test method name ''{0}'' is not valid. Each segment must start with a lowercase letter, min 2 chars, avoid single lowercase letter followed by uppercase.

google.method.name.underscore.regular=Method name ''{0}'' has invalid underscore usage, underscores only allowed between adjacent digits.

google.method.name.underscore.test=Test method name ''{0}'' has invalid underscore usage, underscores only allowed between letters or between digits.
```

Zopsss replied:
@vivek-0509 sorry for the late review, the messages looks good. Please proceed with them

vivek-0509 replied:
done

vivek-0509 committed:
Issue #18420: Add GoogleMethodNameCheck for Google Java Style Guide method naming

vivek-0509 committed:
Issue #17841: Add GoogleMethodName module to google_checks.xml

vivek-0509 commented:
@romani @Zopsss The pr is ready for review. All the comments are addressed and the updates are split into two commits Test methods using underscores to separate logical components are no longer incorrectly flagged as violations
Here are some references from the diff report

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/0f40797_2026173917/reports/diff/apache-ant/index.html#A11

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/0f40797_2026173917/reports/diff/apache-struts/index.html#A153

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/0f40797_2026173917/reports/diff/checkstyle/index.html#A907x

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/0f40797_2026173917/reports/diff/elasticsearch/index.html#A86

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/0f40797_2026173917/reports/diff/spring-integration/index.html#A30

vivek-0509 commented:
@Zopsss No they do not allow special characters only ASCII letters, digits and underscores are allowed for method names,
> Identifiers use only ASCII letters and digits, and, in a small number of cases noted below, underscores. 

https://google.github.io/styleguide/javaguide.html#s5.1-identifier-names

Zopsss commented:
Method [testIsAllowed_emptyAllowedAndDisallowed](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/0f40797_2026173917/reports/diff/apache-struts/index.html#A137) is producing this violation: 

> Method name 'testIsAllowed_emptyAllowedAndDisallowed' has invalid underscore usage, underscores only allowed between adjacent digits.

But [publicPojoDepthOne_sqrBracket](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/0f40797_2026173917/reports/diff/apache-struts/index.html#A140) is not giving any violation. Is this a false-negative? 

---

NVM, the `publicPojoDepthOne_sqrBracket` is a test method so underscores are allowed. My bad.

Zopsss commented:
We have 4 violation messages but the example contains only 3: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/efae5fc_20260117181355/checks/naming/googlemethodname.html#Example1-code

And they do not use the whole violation messages either. Please cover all messages example and do not trim them. Suppress the LineLength violation if necessary. @vivek-0509

Zopsss commented:
Sure, this will work as well. Please modify all the messages to include this.

vivek-0509 commented:
Hey @Zopsss, I have updated the violation messages, for :

- Test input files i used the relevant part of the violation message.
- Example1.java i used the full violation messages as suggested, now covering all 4 message types

I did not wrap the violation messages in Example1 since the relevant part is visible without scrolling for most cases.
Please let me know if any further changes are needed.

Zopsss commented:
GitHub, generate report

vivek-0509 commented:
I have generated the Diff report in comparison to the Method Name Check which was used earlier in google checks to clearly see new and removed violation .

Zopsss reviewed (approved):
Sorry for the delay, the diff report and documentation looks good! Thanks a lot for all the hard work!!!!

romani commented:
With such massive diff report, this update will be a bomb.
And we can can not reliably verify that there is no regression, false positives.

If will be awesome to find some project that uses Google style, and send them updates to test their reaction, before we pour this on heads of whole world.

Is there any ideas?

Maybe comunda?

romani commented:
>Do you know about any other project that enforces google_checks.xml ?

No, users try to customize a bit config, and  I understand it, demand for javadoc on all is too expensive.
Majority immediately remove all about javadoc. Javadoc is applicable only for libraries, so it is very low percentage from general java code base

vivek-0509 commented:
@romani 
i think we need to update the violation message also to not confuse users

`Test method name ''{0}'' is not valid. Each segment must start lowercase, contain at least one uppercase letter for multi-segment names, min 2 chars, avoid single lowercase letter followed by uppercase, and contain only letters, digits, or underscores.`

does this work ??

romani commented:
@lrozenblyum , I know you use Google style.
Can you do us favor to provide feedback on this Check implementation?
We can release it, without activation in Google style config and we can probably us you to test it of your codebase.
If you confirm that no bad side effects appears, we can update config and be more confident in success.

vivek-0509 commented:
GitHub, generate report

vivek-0509 commented:
@romani i have reviewed the diff report, snake case names are now correctly flagged as violations. Will rebase the branch since new check is merged so we need to update the XmlMetaReaderTest

lrozenblyum commented:
> @lrozenblyum , I know you use Google style. Can you do us favor to provide feedback on this Check implementation? We can release it, without activation in Google style config and we can probably us you to test it of your codebase. If you confirm that no bad side effects appears, we can update config and be more confident in success.

Hello @romani. Yes I think I can support this. Does that mean I need to use some nightly build of Checkstyle or will there be a stable release with this rule?

vivek-0509 commented:
Updated the version to 13.4.0, i was getting the spell check ci failure with the diff to add `args `, but my PR does not add args anywhere, i have added to satisfy the CI

vivek-0509 commented:
@Zopsss @romani the PR is related to Google Style project..
Closing this as discussed with Mahfouz
@Atharv3221 You can reuse the work since the PR is in very good state just few tweaks and ready to merge
